androidenhancedvideoplayer/src/main/java/com/profusion/androidenhancedvideoplayer/utils/TrackUtils.kt

```diff
@@ -0,0 +1,89 @@
+package com.profusion.androidenhancedvideoplayer.utils
+
+import androidx.annotation.OptIn
+import androidx.media3.common.C
+import androidx.media3.common.Format
+import androidx.media3.common.TrackGroup
+import androidx.media3.common.TrackSelectionOverride
+import androidx.media3.common.util.UnstableApi
+import androidx.media3.exoplayer.trackselection.DefaultTrackSelector
+import androidx.media3.exoplayer.trackselection.TrackSelector
+
+@OptIn(UnstableApi::class)
+fun DefaultTrackSelector.getAutoTrack(label: String? = null): TrackQualityItem {
+    return TrackQualityItem(
+        builder = buildUponParameters()
+            .clearOverridesOfType(C.TRACK_TYPE_VIDEO)
+            .setAllowMultipleAdaptiveSelections(true),
+        autoLabel = label
+    )
+}
+
+fun generateTrackQualityOptions(trackSelector: TrackSelector? = null): List<TrackQualityItem> {
+    return if (trackSelector is DefaultTrackSelector) {
+        trackSelector.generateTrackQualityOptions()
+    } else {
+        listOf()
+    }
+}
+
+@OptIn(UnstableApi::class)
+fun DefaultTrackSelector.generateTrackQualityOptions():
+    List<TrackQualityItem> {
+    val qualityOptions = ArrayList<TrackQualityItem>()
+
+    val videoQualityOptions = getAvailableVideoTracks().map { (trackIndex, trackGroup) ->
+        TrackQualityItem(
+            format = trackGroup.getFormat(trackIndex),
+            builder = buildUponParameters()
+                .clearOverridesOfType(C.TRACK_TYPE_VIDEO)
+                .addOverride(
+                    TrackSelectionOverride(trackGroup, listOf(trackIndex))
+                )
+        )
+    }.sortedDescending()
+
+    qualityOptions += getAutoTrack()
+    qualityOptions += videoQualityOptions
+
+    return qualityOptions
+}
+
+@OptIn(UnstableApi::class)
+fun DefaultTrackSelector.getAvailableVideoTracks(): ArrayList<Pair<Int, TrackGroup>> {
+    val availableTracks = ArrayList<Pair<Int, TrackGroup>>()
+    val renderTrack = this.currentMappedTrackInfo ?: return availableTracks
+
+    for (rendererIndex in 0 until renderTrack.rendererCount) {
+        if (renderTrack.getRendererType(rendererIndex) != C.TRACK_TYPE_VIDEO) {
+            continue
+        }
+        val trackGroups = renderTrack.getTrackGroups(rendererIndex)
+        for (groupIndex in 0 until trackGroups.length) {
+            for (trackIndex in 0 until trackGroups[groupIndex].length) {
+                if (renderTrack.getTrackSupport(rendererIndex, groupIndex, trackIndex)
+                    != C.FORMAT_HANDLED
+                ) {
+                    continue
+                }
+                if (!trackGroups[groupIndex].getFormat(trackIndex).isVideoTrackPlayable()) {
+                    continue
+                }
+
+                availableTracks += trackIndex to trackGroups[groupIndex]
```

Comment:
Why not create a TrackQualityItem directly? This way you avoid iterating over the items again